pkg/volume/storageos/storageos.go

```diff
@@ -0,0 +1,574 @@
+			volNamespace:    source.VolumeNamespace,
+			pool:            source.Pool,
+			manager:         manager,
+			mounter:         mounter,
+			plugin:          plugin,
+			MetricsProvider: volume.NewMetricsStatFS(getPath(pod.UID, source.VolumeNamespace, source.VolumeName, spec.Name(), plugin.host)),
+		},
+		fsType:      source.FSType,
+		readOnly:    readOnly,
+		devicePath:  storageosDevicePath,
+		diskMounter: &mount.SafeFormatAndMount{Interface: mounter, Runner: exec.New()},
+	}, nil
+}
+
+func (plugin *storageosPlugin) NewUnmounter(pvName string, podUID types.UID) (volume.Unmounter, error) {
+	return plugin.newUnmounterInternal(pvName, podUID, &storageosUtil{}, plugin.host.GetMounter())
+}
+
+func (plugin *storageosPlugin) newUnmounterInternal(pvName string, podUID types.UID, manager storageosManager, mounter mount.Interface) (volume.Unmounter, error) {
+
+	// Parse volume namespace & name from mountpoint if mounted
+	namespace, volName, err := getVolumeInfo(pvName, podUID, plugin.host)
+	if err != nil {
+		return &storageosUnmounter{storageos: &storageos{}}, nil
+	}
+
+	return &storageosUnmounter{
+		storageos: &storageos{
+			podUID:          podUID,
+			pvName:          pvName,
+			volName:         volName,
+			volNamespace:    namespace,
+			manager:         manager,
+			mounter:         mounter,
+			plugin:          plugin,
+			MetricsProvider: volume.NewMetricsStatFS(getPath(podUID, namespace, volName, pvName, plugin.host)),
+		},
+	}, nil
+}
+
+func (plugin *storageosPlugin) NewDeleter(spec *volume.Spec) (volume.Deleter, error) {
+	return plugin.newDeleterInternal(spec, &storageosUtil{})
+}
+
+func (plugin *storageosPlugin) newDeleterInternal(spec *volume.Spec, manager storageosManager) (volume.Deleter, error) {
+	if spec.PersistentVolume != nil && spec.PersistentVolume.Spec.StorageOS == nil {
+		return nil, fmt.Errorf("spec.PersistentVolumeSource.StorageOS is nil")
+	}
+	return &storageosDeleter{
+		storageos: &storageos{
+			pvName:       spec.Name(),
+			volName:      spec.PersistentVolume.Spec.StorageOS.VolumeName,
+			volNamespace: spec.PersistentVolume.Spec.StorageOS.VolumeNamespace,
+			manager:      manager,
+			plugin:       plugin,
+		}}, nil
+}
+
+func (plugin *storageosPlugin) NewProvisioner(options volume.VolumeOptions) (volume.Provisioner, error) {
+	return plugin.newProvisionerInternal(options, &storageosUtil{})
+}
+
+func (plugin *storageosPlugin) newProvisionerInternal(options volume.VolumeOptions, manager storageosManager) (volume.Provisioner, error) {
+	return &storageosProvisioner{
+		storageos: &storageos{
+			manager: manager,
+			plugin:  plugin,
+		},
+		options: options,
+	}, nil
+}
+
+func (plugin *storageosPlugin) ConstructVolumeSpec(volumeName, mountPath string) (*volume.Spec, error) {
+	volNamespace, volName, err := getVolumeFromRef(volumeName)
+	if err != nil {
+		volNamespace = defaultNamespace
+		volName = volumeName
+	}
+	storageosVolume := &v1.Volume{
+		Name: volumeName,
```

Comment:
I'm not sure if this is prone to the volume reconstruction issue. @chakri-nelluri, can you check?